### ISSUE TITLE
[Feature request]: Remember last visited folders/files accross clients?

### ISSUE THREAD

**Issue by CabalCrow** [gokcehan/lf#1463]
lf remembers the last visited files in each folder within a single client (ex: if I entered /usr/share/ folder & then go somewhere else, when I return to the /usr/ folder the cursor would already be on the share folder, rather than the first folder (bin)). Is there a way for lf to preserve the last visited files when a new client is launched?

**Comment by CatsDeservePets** [comment]
Hello @CabalCrow, hello @vagos!

This is a really quick implementation you can paste into your config (might need some fine-tuning):

```sh
# restore last file on start-up
cmd on-init ${{
    last="$HOME/.local/share/lf-last-file"
    if [ -f "$last" ]; then
        lf -remote "send $id select \"$(cat "$last")\""
    fi
}}

# remember current file when closing
cmd on-quit ${{
    if [ -n "$f" ]; then
        printf '%s\n' "$f" > "$HOME/.local/share/lf-last-file"
    fi
}}
```

**Comment by CabalCrow** [comment]
Hello @CatsDeservePets,

The issue in question has to do with something else (though it is similar). Your code restores the last file that was open in the last session. However what this issue was about was a way to preserve "visited" information. The point is as @vagos mentioned when you have to keep opening the same folders in lf it keeps you looking at the files from the last session.

**Comment by CatsDeservePets** [comment]
Oh I see.

Well, you could write your own registry system (some key-value store) using `pre-cd`, `on-cd` and perhaps `on-load`. Before changing the directory, you would have to store `$f` as the value for the current path. After changing, you would have to select the value for the current dir.

Probably possible but tricky to implement.

Official support would have to be implemented in an similar way. We would have to store a giant `map` with all dir names ever visited (perhaps capped at 1000 entries or something like that).

**Comment by CabalCrow** [comment]
The thing is lf already does this. If you don't close lf and you move around it stores and remember the last visited file in all folders. Only when you close it this information is lost - so I assume it is saved someone in ram, but not on disk. My guess is the ``map`` is already in code, so all that is left to do is to find where it is, and then (de)serialize it.

That being said I don't have go coding experience so I could be wrong.

**Comment by CatsDeservePets** [comment]
Nah, it works differently. In `lf`, every loaded directory is its own object of type
```go
type dir struct {
	loading      bool       // directory is loading from disk
	loadTime     time.Time  // last load time
	ind          int        // index of current entry in files
	pos          int        // position of current entry in ui
...
}
```
which happens to have a property `pos`.

**Comment by CabalCrow** [comment]
So can all of this be saved in a file, where you have ``dirname-pos`` of all the visited files and then you load from this file?

**Comment by CatsDeservePets** [comment]
Theoretically speaking, yes. The question is, how simple would it be to implement and would it be worth it.
Kindly asking @joelim-work to give his thoughts.

**Comment by joelim-work** [comment]
> If you don't close lf and you move around it stores and remember the last visited file in all folders. Only when you close it this information is lost - so I assume it is saved someone in ram, but not on disk.

That is correct, the internal state is managed using a data structure, but this only exists in application memory and not persisted to disk, so the information is lost after you quit.

Each directory is represented as a struct with some fields:

```go
type dir struct {
	files        []*file    // displayed files in directory including or excluding hidden ones
	ind          int        // index of current entry in files
	// ...
}
```

This is somewhat simplified, but you can think of `dir.files` as an array containing file objects, and `dir.ind` as the index showing which file the cursor is on.

When it comes to application programming, a lot of things are possible - they just have to be programmed. In other words this is basically a feature request to persist part of the application state (some other parts are already persisted like the history, clipboard, marks, tags, etc.).

One thing I should point out is that it might not make sense to persist a number like `dir.ind`, if some files gets created/deleted while `lf` is not running then this can mess up the order of the files.

**Comment by CatsDeservePets** [comment]
After thinking more about this, I will close this as `not planned`. It seems rather complex to implement and too niche of a feature. Sorry!
